mailpoet/lib/WooCommerce/CouponPreProcessor.php

```diff
@@ -0,0 +1,60 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\WooCommerce;
+
+use MailPoet\Entities\NewsletterEntity;
+use MailPoet\Newsletter\Renderer\Blocks\Coupon;
+
+class CouponPreProcessor {
+  public function processCoupons(NewsletterEntity $newsletter, array $block, bool $preview = false): array {
+    if ($preview) {
+      return $block;
+    }
+
+    $couponBlocks = $this->findCouponBlocks($block);
+    foreach ($couponBlocks as $couponBlock) {
+      if (empty($couponBlock['couponId'])) {
+        $couponBlock['couponId'] = $this->generateCoupon();
+      }
+    }
+
+    return $block;
+  }
+
+  private function findCouponBlocks(array $block): array {
+    $coupons = [];
+
+    foreach ($block['blocks'] as $innerBlock) {
+      if (isset($innerBlock['blocks']) && !empty($innerBlock['blocks'])) {
+        $coupons = array_merge($coupons, $this->findCouponBlocks($innerBlock));
+      }
+      if ($innerBlock['type'] === Coupon::TYPE) {
+        $coupons[] = $innerBlock;
+      }
+    }
+
+    return $coupons;
+  }
+
+  private function generateCoupon(string $couponCode = null): int {
+    $coupon = new \WC_Coupon();
+    $code = $couponCode ?? $this->generateRandomCode();
+    $coupon->set_code($code);
+
+    return $coupon->save();
```

Comment:
What happens if we accidentally generate an existing code? Will `coupon->save` throw an exception or return an error? I just want to ensure we don't end up with an error in preprocessing that would block sending.

From what I experienced when testing, when we set coupon type percent and coupon value more than 100, there will be an error.

Reply:
That's correct, it'll throw, I was under the impression to leave this to the validation ticket, but your `it won't get released under a feature flag` is valid and I'll manage this.